File: src/grid/geodesic_grid.py
"""
Geodesic icosphere grid built by recursive midpoint subdivision.

Level-6 produces 40,962 cells (vertices of the subdivided icosahedron,
projected onto the unit sphere).

Cell indices 0..11 are the 12 original icosahedron vertices (pentagonal cells
with 5 neighbours). Indices 12..N-1 are hexagonal cells with 6 neighbours.
"""

import math
import numpy as np
from scipy.spatial import cKDTree
from typing import List, Tuple


class GeodesicGrid:
    """Geodesic icosphere grid on the unit sphere."""

    def __init__(self, level: int = 6) -> None:
        """
        Construct a geodesic grid via recursive subdivision.

        Args:
            level: Number of subdivision passes. Level 6 gives 40,962 cells.
        """
        self._level = level
        self._vertices: np.ndarray = np.zeros((0, 3), dtype=np.float64)
        self._faces: List[Tuple[int, int, int]] = []
        self._adjacency: List[List[int]] = []
        self._kdtree: cKDTree = None

        self._build_icosahedron()
        for _ in range(level):
            self._subdivide()
        self._project_to_sphere()
        self._build_adjacency()
        self._build_kdtree()

    # --- Public properties ---

    @property
    def cell_count(self) -> int:
        """Total number of cells in the grid."""
        return len(self._vertices)

    # --- Public methods ---

    def cell_center(self, cell_idx: int) -> Tuple[float, float]:
        """
        Get cell center coordinates.

        Args:
            cell_idx: Cell index (0 to cell_count-1)

        Returns:
            (latitude, longitude) in degrees
        """
        return self.cell_to_latlon(cell_idx)

    def cell_neighbours(self, cell_idx: int) -> List[int]:
        """
        Get indices of cells neighbouring the given cell.

        Args:
            cell_idx: Cell index

        Returns:
            List of 5 or 6 neighbour indices, sorted ascending
        """
        return self._adjacency[cell_idx]

    def latlon_to_cell(self, lat: float, lon: float) -> int:
        """
        Find the cell at the given geographic location.

        Args:
            lat: Latitude in degrees (-90 to +90)
            lon: Longitude in degrees (-180 to +180)

        Returns:
            Cell index of the nearest cell
        """
        xyz = self._latlon_to_xyz(lat, lon)
        _, idx = self._kdtree.query(xyz)
        return int(idx)

    def latlon_to_cells_batch(self, latlons: np.ndarray) -> np.ndarray:
        """
        Find cells for multiple geographic locations (vectorized).

        Args:
            latlons: (N, 2) array of [lat, lon] in degrees

        Returns:
            (N,) array of cell indices
        """
        lat_r = np.radians(latlons[:, 0])
        lon_r = np.radians(latlons[:, 1])
        xyz = np.column_stack([
            np.cos(lat_r) * np.cos(lon_r),
            np.cos(lat_r) * np.sin(lon_r),
            np.sin(lat_r)
        ])
        _, indices = self._kdtree.query(xyz)
        return indices.astype(np.int32)

    def cell_to_latlon(self, cell_idx: int) -> Tuple[float, float]:
        """
        Get geographic coordinates of a cell center.

        Args:
            cell_idx: Cell index

        Returns:
            (latitude, longitude) in degrees
        """
        x, y, z = self._vertices[cell_idx]
        return self._xyz_to_latlon(x, y, z)

    # --- Private build methods ---

    def _build_icosahedron(self) -> None:
        """Build the 12-vertex icosahedron via golden ratio construction."""
        phi = (1.0 + math.sqrt(5.0)) / 2.0

        # Canonical vertex ordering (must match C++ implementation)
        vertices_raw = [
            (0, 1, phi),
            (0, -1, phi),
            (0, 1, -phi),
            (0, -1, -phi),
            (1, phi, 0),
            (-1, phi, 0),
            (1, -phi, 0),
            (-1, -phi, 0),
            (phi, 0, 1),
            (-phi, 0, 1),
            (phi, 0, -1),
            (-phi, 0, -1),
        ]

        self._vertices = np.array([
            self._normalize_vec(np.array(v, dtype=np.float64))
            for v in vertices_raw
        ], dtype=np.float64)

        # 20 faces, CCW from outside (computed from golden ratio icosahedron)
        # Each edge appears in exactly 2 faces; each vertex has degree 5 or 6
        self._faces = [
            (0, 1, 8), (0, 9, 1), (0, 4, 5), (0, 8, 4), (0, 5, 9),
            (1, 7, 6), (1, 6, 8), (1, 9, 7), (2, 10, 3), (2, 3, 11),
            (2, 5, 4), (2, 4, 10), (2, 11, 5), (3, 6, 7), (3, 10, 6),
            (3, 7, 11), (4, 8, 10), (5, 11, 9), (6, 10, 8), (7, 9, 11),
        ]

    def _subdivide(self) -> None:
        """Perform one pass of midpoint subdivision."""
        midpoint_cache = {}

        def get_or_create_midpoint(a: int, b: int) -> int:
            """Get or create midpoint vertex index."""
            key = (min(a, b), max(a, b))
            if key in midpoint_cache:
                return midpoint_cache[key]

            # Compute midpoint (not yet normalized)
            mid = (self._vertices[a] + self._vertices[b]) * 0.5
            new_idx = len(self._vertices)
            self._vertices = np.vstack([self._vertices, [mid]])
            midpoint_cache[key] = new_idx
            return new_idx

        new_faces = []
        for a, b, c in self._faces:
            mab = get_or_create_midpoint(a, b)
            mbc = get_or_create_midpoint(b, c)
            mca = get_or_create_midpoint(c, a)

            new_faces.extend([
                (a, mab, mca),
                (b, mbc, mab),
                (c, mca, mbc),
                (mab, mbc, mca),
            ])

        self._faces = new_faces

    def _project_to_sphere(self) -> None:
        """Project all vertices onto the unit sphere."""
        norms = np.linalg.norm(self._vertices, axis=1, keepdims=True)
        self._vertices = self._vertices / norms

    def _build_adjacency(self) -> None:
        """Build neighbour lists from face topology."""
        N = len(self._vertices)
        adjacency_sets = [set() for _ in range(N)]

        for a, b, c in self._faces:
            adjacency_sets[a].add(b)
            adjacency_sets[a].add(c)
            adjacency_sets[b].add(a)
            adjacency_sets[b].add(c)
            adjacency_sets[c].add(a)
            adjacency_sets[c].add(b)

        self._adjacency = [sorted(s) for s in adjacency_sets]

    def _build_kdtree(self) -> None:
        """Build KD-tree spatial index."""
        self._kdtree = cKDTree(self._vertices)

    # --- Private helpers ---

    @staticmethod
    def _normalize_vec(v: np.ndarray) -> np.ndarray:
        """Normalize a 3D vector to unit length."""
        return v / np.linalg.norm(v)

    @staticmethod
    def _latlon_to_xyz(lat: float, lon: float) -> np.ndarray:
        """Convert lat/lon (degrees) to unit-sphere XYZ."""
        lat_r = math.radians(lat)
        lon_r = math.radians(lon)
        return np.array([
            math.cos(lat_r) * math.cos(lon_r),
            math.cos(lat_r) * math.sin(lon_r),
            math.sin(lat_r)
        ], dtype=np.float64)

    @staticmethod
    def _xyz_to_latlon(x: float, y: float, z: float) -> Tuple[float, float]:
        """Convert unit-sphere XYZ to lat/lon (degrees)."""
        # Guard against floating-point rounding
        z = max(-1.0, min(1.0, z))
        lat = math.degrees(math.asin(z))
        lon = math.degrees(math.atan2(y, x))
        return (lat, lon)

File: src/simulation/grid/geodesic_grid.cpp
#include "geodesic_grid.h"
#include <algorithm>
#include <limits>
#include <unordered_set>

namespace geoforge {

GeodesicGrid::GeodesicGrid(int level)
    : level_(level),
      hash_cell_size_(2.0f / kHashRes) {
    BuildIcosahedron();
    for (int i = 0; i < level; ++i) {
        SubdivideOnce();
    }
    ProjectToSphere();
    BuildAdjacency();
    BuildSpatialHash();
}

std::pair<float, float> GeodesicGrid::CellCenter(int cell_idx) const {
    return CellToLatLon(cell_idx);
}

const std::vector<int>& GeodesicGrid::CellNeighbours(int cell_idx) const {
    return adjacency_[cell_idx];
}

int GeodesicGrid::LatLonToCell(float lat, float lon) const {
    Vec3 xyz = LatLonToXyz(lat, lon);
    return FindNearest(xyz.x, xyz.y, xyz.z);
}

std::pair<float, float> GeodesicGrid::CellToLatLon(int cell_idx) const {
    const Vec3& v = vertices_[cell_idx];
    return XyzToLatLon(v.x, v.y, v.z);
}

// --- Private build methods ---

void GeodesicGrid::BuildIcosahedron() {
    const float phi = (1.0f + std::sqrt(5.0f)) / 2.0f;

    // Canonical vertex ordering (must match Python implementation)
    std::vector<Vec3> vertices_raw = {
        Vec3(0, 1, phi).Normalized(),
        Vec3(0, -1, phi).Normalized(),
        Vec3(0, 1, -phi).Normalized(),
        Vec3(0, -1, -phi).Normalized(),
        Vec3(1, phi, 0).Normalized(),
        Vec3(-1, phi, 0).Normalized(),
        Vec3(1, -phi, 0).Normalized(),
        Vec3(-1, -phi, 0).Normalized(),
        Vec3(phi, 0, 1).Normalized(),
        Vec3(-phi, 0, 1).Normalized(),
        Vec3(phi, 0, -1).Normalized(),
        Vec3(-phi, 0, -1).Normalized(),
    };

    vertices_ = vertices_raw;

    // 20 faces, CCW from outside (computed from golden ratio icosahedron)
    faces_ = {
        {0, 1, 8}, {0, 9, 1}, {0, 4, 5}, {0, 8, 4}, {0, 5, 9},
        {1, 7, 6}, {1, 6, 8}, {1, 9, 7}, {2, 10, 3}, {2, 3, 11},
        {2, 5, 4}, {2, 4, 10}, {2, 11, 5}, {3, 6, 7}, {3, 10, 6},
        {3, 7, 11}, {4, 8, 10}, {5, 11, 9}, {6, 10, 8}, {7, 9, 11},
    };
}

void GeodesicGrid::SubdivideOnce() {
    std::unordered_map<uint64_t, int> midpoint_cache;

    auto EdgeKey = [](int a, int b) -> uint64_t {
        if (a > b) std::swap(a, b);
        return (static_cast<uint64_t>(a) << 32) | static_cast<uint64_t>(b);
    };

    auto GetMidpoint = [&](int a, int b) -> int {
        uint64_t key = EdgeKey(a, b);
        auto it = midpoint_cache.find(key);
        if (it != midpoint_cache.end()) return it->second;

        // Average, not yet normalized
        Vec3 mid = (vertices_[a] + vertices_[b]) * 0.5f;
        int idx = static_cast<int>(vertices_.size());
        vertices_.push_back(mid);
        midpoint_cache[key] = idx;
        return idx;
    };

    std::vector<std::array<int, 3>> new_faces;
    new_faces.reserve(faces_.size() * 4);

    for (auto& f : faces_) {
        int mab = GetMidpoint(f[0], f[1]);
        int mbc = GetMidpoint(f[1], f[2]);
        int mca = GetMidpoint(f[2], f[0]);

        new_faces.push_back({f[0], mab, mca});
        new_faces.push_back({f[1], mbc, mab});
        new_faces.push_back({f[2], mca, mbc});
        new_faces.push_back({mab, mbc, mca});
    }

    faces_ = std::move(new_faces);
}

void GeodesicGrid::ProjectToSphere() {
    for (auto& v : vertices_) {
        v = v.Normalized();
    }
}

void GeodesicGrid::BuildAdjacency() {
    int N = static_cast<int>(vertices_.size());
    adjacency_.assign(N, {});

    // Temporary sets to deduplicate
    std::vector<std::unordered_set<int>> adj_sets(N);

    for (auto& f : faces_) {
        int a = f[0], b = f[1], c = f[2];
        adj_sets[a].insert(b);
        adj_sets[a].insert(c);
        adj_sets[b].insert(a);
        adj_sets[b].insert(c);
        adj_sets[c].insert(a);
        adj_sets[c].insert(b);
    }

    for (int i = 0; i < N; ++i) {
        adjacency_[i].assign(adj_sets[i].begin(), adj_sets[i].end());
        std::sort(adjacency_[i].begin(), adjacency_[i].end());
    }

    // Clear faces to free memory
    faces_.clear();
    faces_.shrink_to_fit();
}

void GeodesicGrid::BuildSpatialHash() {
    hash_buckets_.resize(kHashRes * kHashRes * kHashRes);

    for (int i = 0; i < static_cast<int>(vertices_.size()); ++i) {
        auto bucket = HashBucket(vertices_[i].x, vertices_[i].y, vertices_[i].z);
        int flat = bucket[0] * kHashRes * kHashRes
                 + bucket[1] * kHashRes
                 + bucket[2];
        hash_buckets_[flat].push_back(i);
    }
}

// --- Spatial hash helpers ---

std::array<int, 3> GeodesicGrid::HashBucket(float x, float y, float z) const {
    auto Clamp = [](int v, int lo, int hi) {
        return v < lo ? lo : v > hi ? hi : v;
    };

    int bx = Clamp(static_cast<int>((x + 1.0f) / hash_cell_size_), 0, kHashRes - 1);
    int by = Clamp(static_cast<int>((y + 1.0f) / hash_cell_size_), 0, kHashRes - 1);
    int bz = Clamp(static_cast<int>((z + 1.0f) / hash_cell_size_), 0, kHashRes - 1);
    return {bx, by, bz};
}

int GeodesicGrid::FindNearest(float x, float y, float z) const {
    auto bucket = HashBucket(x, y, z);
    int best_idx = -1;
    float best_dist_sq = std::numeric_limits<float>::max();

    // Probe expanding shells (up to 4 ensures all cells are found even with rounding)
    for (int shell = 0; shell <= 4; ++shell) {
        for (int dx = -shell; dx <= shell; ++dx) {
        for (int dy = -shell; dy <= shell; ++dy) {
        for (int dz = -shell; dz <= shell; ++dz) {
            // Only visit cells on the surface of the current shell
            if (shell > 0 &&
                std::abs(dx) < shell &&
                std::abs(dy) < shell &&
                std::abs(dz) < shell) {
                continue;
            }

            int nx = bucket[0] + dx;
            int ny = bucket[1] + dy;
            int nz = bucket[2] + dz;
            if (nx < 0 || nx >= kHashRes) continue;
            if (ny < 0 || ny >= kHashRes) continue;
            if (nz < 0 || nz >= kHashRes) continue;

            int flat = nx * kHashRes * kHashRes + ny * kHashRes + nz;
            for (int idx : hash_buckets_[flat]) {
                const Vec3& v = vertices_[idx];
                float d2 = (v.x-x)*(v.x-x) + (v.y-y)*(v.y-y) + (v.z-z)*(v.z-z);
                if (d2 < best_dist_sq) {
                    best_dist_sq = d2;
                    best_idx = idx;
                }
            }
        }}}

        if (best_idx >= 0) {
            // Early termination: if best distance < shell boundary, no closer cell exists
            float shell_boundary = (shell * hash_cell_size_) * (shell * hash_cell_size_);
            if (best_dist_sq <= shell_boundary) break;
        }
    }

    return best_idx;
}

// --- Conversion helpers ---

GeodesicGrid::Vec3 GeodesicGrid::LatLonToXyz(float lat_deg, float lon_deg) {
    float lat_r = lat_deg * (3.14159265359f / 180.0f);
    float lon_r = lon_deg * (3.14159265359f / 180.0f);
    return Vec3(
        std::cos(lat_r) * std::cos(lon_r),
        std::cos(lat_r) * std::sin(lon_r),
        std::sin(lat_r)
    );
}

std::pair<float, float> GeodesicGrid::XyzToLatLon(float x, float y, float z) {
    // Guard against floating-point rounding
    z = std::clamp(z, -1.0f, 1.0f);
    float lat = std::asin(z) * (180.0f / 3.14159265359f);
    float lon = std::atan2(y, x) * (180.0f / 3.14159265359f);
    return {lat, lon};
}

}  // namespace geoforge
